bakery/service/update_generation.go

```diff
@@ -0,0 +1,187 @@
+		keyValuePair := strings.Split(replacement, propertyKeyValueSeparator)
+		if len(keyValuePair) != 2 {
+			fmt.Printf("Unable to parse provided replacement: %s - expecting key and value to be separated with '%s'", replacement, propertyKeyValueSeparator)
+			continue
+		}
+		replacer = replacer.add(keyValuePair[0], keyValuePair[1])
+	}
+	return replacer
+}
+
+func generateAncestors(previousImageName string, dependenciesMap map[string][]*DockerImage, replacer nameReplacer, recursive bool, skipExistingDirectories bool) error {
+	ancestorsToGenerate, err := getChildrenToUpdate(previousImageName, dependenciesMap, replacer, recursive)
+	fmt.Println("Found ancestors to recreate:")
+	for _, ancestor := range ancestorsToGenerate {
+		fmt.Println(ancestor.childOriginImageName)
+	}
+	if err != nil {
+		return err
+	}
+	for _, ancestor := range ancestorsToGenerate {
+		err := generateChildProjects(ancestor, replacer, skipExistingDirectories)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func generateChildProjects(ancestor imageToGenerate, replacer nameReplacer, skipExistingDirectories bool) error {
+	targetPath := ancestor.newImageDirectory
+	exists, err := checkDirectoryExistence(targetPath, skipExistingDirectories)
+	if err != nil {
+		return err
+	}
+
+	if exists || ancestor.childOriginDirectory == targetPath { // when no explicit version in image name, just update the file inside
+		err = updateInPlace(ancestor, replacer)
+	} else {
+		err = copyImageDirectory(ancestor.childOriginDirectory, targetPath)
+	}
+	if err != nil {
+		return err
+	}
+	return updateImageBaseInformation(targetPath, ancestor.newParentImage, ancestor.originalParentImage)
+}
+
+func updateImageBaseInformation(path string, newParentImageName string, originalParentName string) error {
+	dockerFilePath := filepath.Join(path, dockerFileTemplateName)
+	read, err := ioutil.ReadFile(dockerFilePath)
+	if err != nil {
+		return err
+	}
+
+	newContents := strings.Replace(string(read), originalParentName, newParentImageName, -1)
+	newContents = strings.Replace(newContents, dynamicImageVersionName(originalParentName), dynamicImageVersionName(newParentImageName), -1)
+
+	return ioutil.WriteFile(dockerFilePath, []byte(newContents), 0)
+}
+
+func copyImageDirectory(childOriginDirectory string, targetPath string) error {
+	copyCommand := exec.Command("cp", "-r", childOriginDirectory, targetPath)
+	copyCommand.Stdin = os.Stdin
+	copyCommand.Stdout = os.Stdout
+	copyCommand.Stderr = os.Stderr
+	return copyCommand.Run()
+
+}
+
+func simpleNameReplacer(oldName, newName string) func(string) string {
+	return func(s string) string {
+		return strings.ReplaceAll(s, oldName, newName)
+	}
+}
+
+type nameReplacer struct {
+	replacements []changeNameFn
+}
+
+func (n nameReplacer) add(from, to string) nameReplacer {
+	n.replacements = append(n.replacements, simpleNameReplacer(from, to))
+	from2 := strings.ToUpper(strings.Replace(from, "-", "_", -1))
```

Comment:
Could be extracted to separate function with meaningful name (e.g. `buildImageVersionVariable`).
```
	strings.ToUpper(strings.Replace(from, "-", "_", -1))
```

Reply:
done